test/core/end2end/fixtures/h2_tls_common.h

```diff
@@ -238,8 +238,8 @@ inline grpc_channel_credentials* create_tls_channel_credentials(
 inline grpc_server_credentials* create_tls_server_credentials(
     fullstack_secure_fixture_data* ffd) {
   grpc_tls_credentials_options* options = grpc_tls_credentials_options_create();
-  options->set_min_tls_version(ffd->tls_version);
-  options->set_max_tls_version(ffd->tls_version);
+  grpc_tls_credentials_options_set_min_tls_version(options, ffd->tls_version);
+  grpc_tls_credentials_options_set_max_tls_version(options, ffd->tls_version);
```

Comment:
client set up needs to be changed into using C APIs as well(line ~220)

Reply:
Our original test sets the TLS version as a prerequisite: it assumes every tests should have the same TLS version settings and it should work properly. That's why it's inside the common function.
But now since we are testing the version setting behavior, maybe we could turn that into separate test cases and test some more scenarios? A couple of cases I could think of:

1. client[min 1.2, max 1.3] server [min 1.2, max1.3] -> supposed to work
2. client[min 1.3, max 1.3] server [min 1.3, max1.3] -> supposed to work
3. client[min 1.2, max 1.2] server [min 1.3, max1.3] -> what will happen in this case? I don't remember off hand but you can probably do a quick experiment here
4. client[min 1.3, max 1.3] server [min 1.2, max1.2] -> not sure as well
5. client/server[min 1.3, max 1.2] -> what will happen in this case? initialization failure or runtime failure?

To test these, you probably don't need fancy creds reloading behaviors so you can just go with StaticDataProvider.

Reply:
I will address the rest later, but for tls version mismatches between server/client, I tested locally and it would cause handshake error. This could be tested but the rpc wouldn't be returned immediately and instead it would time out which result in a ~10 seconds unit test duration. Is this acceptable? Or is there other ways to terminate it sooner?

Reply:
Hmmm I see. What exactly is the error message? It times out probably because it is re-trying the handshake. That's probably why core tests usually don't include failure cases. But we shall probably add a few these kind of tests. This was also discussed in another thread I had with @markdroth 
What I am think right now is to create a test without retry(probably some code in `h2_full_no_retry.cc` can be a reference), which hopefully will make things a bit easier. @markdroth Hi Mark, can I know what's the typical way for writing these e2e tests against failure behaviors? Do you have any other suggestions?

Reply:
I am assuming that we're using the default `pick_first` LB policy and that there is only one address returned by the resolver, since I think that should be the case in all of the core e2e tests.

If there is a handshake failure, then the connection attempt will fail, and `pick_first` will put the channel in `TRANSIENT_FAILURE`.  Any RPC sent on the channel will then fail by default.  However, if the RPC is using `wait_for_ready`, then it will remain queued until either (a) a subsequent connection attempt succeeds, or (b) the RPC's deadline expires.  The individual test can control whether it is using `wait_for_ready`, and it can expect the RPC to fail as appropriate.

Retries should not be an issue here.  If the test is not explicitly configuring retries, then there should not be any retries in this case.

Reply:
Thanks Mark for the reply!
If I understand correctly, in our case, the RPC should fail directly, as I didn't find any configurations for `wait_for_ready` in our TLS tests. That seems an unexpected behavior.
@rockspore let's do some investigations offline and see what is the cause of this particular problem. Feel free to let me know if you have any concerns. Thanks!

Reply:
Thanks Zhen and Mark for the details. I realize I was referring to the cpp end2end test I added when I mentioned the timeout. Sorry for misleading.

As for the core end2end test, it's actually not clear to me how to construct a failed test due to handshake errors by adding a fixture, since each fixture will be used to run all the test scenarios, which contain successful and failed RPCs. Looking at the definition of the `grpc_end2end_test_config`, I can't find any existing field which can inform the test that the fixture can cause issues when establishing the channel. Hypothetically, even if we added a new field there, all the test functions invoked in `end2end_tests.cc` would have to be modified to check it, which would be a sizable effort.

Let me know if there is some different approach that I overlooked here. Otherwise, I wonder if we could skip tls version mismatch tests in core end2end and only test it in the cpp ones. @ZhenLian

Reply:
Discussed with Zhen offline. I will need to add new fixtures as well as new tests with some specific `_test_options` in `generate_tests.bzl`. Will come back once I get that done.

Reply:
I added fixtures and test for the bad tls versions. For some reason, the corresponding test couldn't end itself till the 10 seconds timeout specified by at [test_config.cc](https://github.com/grpc/grpc/blob/7c5e3267981b5cdef015438040ad6d5e62ff4044/test/core/util/test_config.cc#L163). I have confirmed that `grpc_shutdown()` in the main function in `h2_tls_wrong_versions.cc` actually returned. To reproduce this, one can run:

```
bazel test //test/core/end2end:h2_tls_wrong_versions_test@status_unavailable --test_output=streamed
```

when this branch is checked out.

@ZhenLian PTAL when you get a chance. Thanks.